packages/opentelemetry-basic-tracer/src/export/BatchSpanProcessor.ts

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { TraceOptions } from '@opentelemetry/types';
+import { SpanProcessor } from '../SpanProcessor';
+import { SpanExporter } from './SpanExporter';
+import { Span } from '../Span';
+import { ReadableSpan } from './ReadableSpan';
+import { unrefTimer } from '../platform';
+import { BufferConfig } from '../types';
+
+const DEFAULT_BUFFER_SIZE = 100;
+const DEFAULT_BUFFER_TIMEOUT = 20000;
+
+/**
+ * Implementation of the {@link SpanProcessor} that batches spans exported by
+ * the SDK then pushes them to the exporter pipeline.
+ */
+export class BatchSpanProcessor implements SpanProcessor {
+  private readonly _bufferSize: number = DEFAULT_BUFFER_SIZE;
+  private readonly _bufferTimeout: number = DEFAULT_BUFFER_TIMEOUT;
+  private _finishedSpans: ReadableSpan[] = [];
+  private _resetTimeout = false;
+  private _bufferTimeoutInProgress = false;
+
+  constructor(private readonly _exporter: SpanExporter, config?: BufferConfig) {
+    if (config) {
+      this._bufferSize = isNaN(Number(config.bufferSize))
+        ? DEFAULT_BUFFER_SIZE
+        : Number(config.bufferSize);
+      this._bufferTimeout = isNaN(Number(config.bufferTimeout))
+        ? DEFAULT_BUFFER_TIMEOUT
+        : Number(config.bufferTimeout);
+    }
+  }
+
+  // does nothing.
+  onStart(span: Span): void {}
+
+  onEnd(span: Span): void {
+    if (span.context().traceOptions !== TraceOptions.SAMPLED) return;
+    this._addToBuffer(span.toReadableSpan());
+  }
+
+  shutdown(): void {
+    this._exporter.shutdown();
+  }
+
+  /** Add a span in the buffer. */
+  private _addToBuffer(span: ReadableSpan) {
+    this._finishedSpans.push(span);
+    if (this._finishedSpans.length > this._bufferSize) {
+      this._flush();
+    }
+
+    if (this._bufferTimeoutInProgress) {
+      this._resetBufferTimeout();
```

Comment:
shouldn't it be race (max buffer size or timeout)? If we reset the buffer always and we add one span at every 19999ms we won't report for 100 * 19999ms = ~33 mins

Reply:
Yeah this could be an annoying issue, i think the buffer timeout should be more like a flush ticker, so we end up with behavior where the tracer flushes every N seconds AND flushes when the buffer is at capacity.

Reply:
+1 to this. Is there a scenario where we would ever want to reset the timer? We would send every N seconds, regardless of when the last span was added to the buffer. If the buffer becomes size M, we flush and `clearTimeout`.

Reply:
Do you suggest to remove the reset functionality? If we do that, every span end we are creating a setting new timer. Let me know if you have better approach to handle this.

Reply:
I have made the changes in https://github.com/open-telemetry/opentelemetry-js/pull/238/commits/8b362fbd0e8d60b94777d089434ed8d0439bb808, ptal